File: Freestylecoding.Math.FSharp.Tests/Natural/UnaryNegationOperators.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public UnaryNegationOperators() =
    static member Negation( n:'T when 'T :> System.Numerics.IUnaryNegationOperators<'T,'T> ) : 'T =
        'T.op_UnaryNegation( n )

    static member CheckedNegation( n:'T when 'T :> System.Numerics.IUnaryNegationOperators<'T,'T> ) : 'T =
        'T.op_CheckedUnaryNegation( n )

    [<Theory>]
    [<InlineData(  0u )>]   // Sanity
    [<InlineData(  1u )>]   // Sanity
    [<InlineData( 42u )>]   // Sanity
    member public this.Sanity value =
        let exc = Record.Exception(
            fun () ->
                UnaryNegationOperators.Negation( Natural( [value] ) )
                |> ignore
        )
        Assert.NotNull( exc )
        Assert.IsType<System.OverflowException>( exc )

    [<Theory>]
    [<InlineData(  0u )>]   // Sanity
    [<InlineData(  1u )>]   // Sanity
    [<InlineData( 42u )>]   // Sanity
    member public this.CheckedSanity value =
        let exc = Record.Exception(
            fun () ->
                UnaryNegationOperators.CheckedNegation( Natural( [value] ) )
                |> ignore
        )
        Assert.NotNull( exc )
        Assert.IsType<System.OverflowException>( exc )

File: Freestylecoding.Math.CSharp.Tests/Natural/UnaryNegationOperators.cs
using System.Numerics;
using Xunit;

namespace Natural.Tests;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class UnaryNegationOperators {
	private static T Negation<T>( T n ) where T : IUnaryNegationOperators<T, T> =>
		-n;
	private static T CheckedNegation<T>( T n ) where T : IUnaryNegationOperators<T, T> =>
		checked(-n);

	[Theory]
	[InlineData( 0u )]   // Sanity
	[InlineData( 1u )]   // Sanity
	[InlineData( 42u )]   // Sanity
	public void Sanity( uint value ) {
		System.Exception exc = Record.Exception(
			() =>
				Negation( new Natural( value ) )
		);

		Assert.NotNull( exc );
		Assert.IsType<System.OverflowException>( exc );
	}

	[Theory]
	[InlineData( 0u )]   // Sanity
	[InlineData( 1u )]   // Sanity
	[InlineData( 42u )]   // Sanity
	public void CheckedSanity( uint value ) {
		System.Exception exc = Record.Exception(
			() =>
				UnaryNegationOperators.CheckedNegation( new Natural( value ) )
		);
		Assert.NotNull( exc );
		Assert.IsType<System.OverflowException>( exc );
	}
}